internal/filter/operators.go

```diff
@@ -0,0 +1,513 @@
+package filter
+
+import (
+	"fmt"
+	datetime "github.com/di-wu/xsd-datetime"
+	"github.com/elimity-com/scim/schema"
+	"github.com/scim2/filter-parser/v2"
+	"strings"
+	"time"
+)
+
+func cmpBool(ref bool, cmp func(v, ref bool) error) func(interface{}) error {
+	return func(i interface{}) error {
+		value, ok := i.(bool)
+		if !ok {
+			return fmt.Errorf("given value is not a boolean: %v", i)
+		}
+		return cmp(value, ref)
+	}
+}
+
+func cmpDataTime(ref time.Time, cmp func(v, ref time.Time) error) func(interface{}) error {
+	return func(i interface{}) error {
+		date, ok := i.(string)
+		if !ok {
+			return fmt.Errorf("given value is not a string: %v", i)
+		}
+		value, err := datetime.Parse(date)
+		if err != nil {
+			return err
+		}
+		return cmp(value, ref)
+	}
+}
+
+func cmpDecimal(ref float64, cmp func(v, ref float64) error) func(interface{}) error {
+	return func(i interface{}) error {
+		value, ok := toFloat(i)
+		if !ok {
+			return fmt.Errorf("given value is not a float: %v", i)
+		}
+		return cmp(value, ref)
+	}
```

Comment:
> Is it okay to just return false in the comparison functions if those conversions fail?

I replaced `return false` with a descriptive error message.

It would be strange if these type conversions failed. This would mean that that the value in the given resource (that needs to pass the filter) does not have the same type in the corresponding field. 

e.g.
```
resource := map[string]interface{}{"userName":x}
```
The conversion would fail if `x` is not a string. But this should be a string in the first place. Any thoughts on this @jdeflander?

Reply:
> Any thoughts on this @jdeflander?

If the type conversions can only fail due to a bug in this library, then using `panic` is the most appropriate IMO.